### ISSUE TITLE
Cmip6MeteorDataGetter might pull training data inconsistently

### ISSUE THREAD

**Issue by normansteinert** [benmsanderson/METEOR#22]
The data getter may pull data CMIP data inconsistently. My suspicion is that this may be related to ESGF server availability. 

I pulled data on different days and got a slightly different models working. In this particular case, some piC data that worked inconsistently, but might also be an issue with scenario data. My guess is that this is model-specific.  

Ideally, you get the data once, and have all training data stored locally in 'cache'. But this might be sensitive to when this is done, and is going to be a problem with reproducibility, and the plot scripts working universally. 

Code:
data_getter = Cmip6MeteorDataGetter(exps=["piControl", "abrupt-4xCO2", "historical", "ssp245"], flds = flds, dbe=['CMIP','CMIP','CMIP', 'ScenarioMIP'])

This is an example error message from 'make_meteor_training_data' for models that don't work:

![Screenshot 2024-11-22 at 09 31 51](https://github.com/user-attachments/assets/940ce152-318d-415e-a0d5-774073be2079)

**Comment by benmsanderson** [comment]
Yes - I have run into this problem before.  I think it's deeply systemic in the way that ESGF relies on distributed resources which are not always available.  What I'd need to check on is whether our current scripts are attempting to grab data from all nodes - or just giving up when the first node has issues.

**Comment by benmsanderson** [comment]
Ignore my prior message - it's not ESGF - we're pulling this data from the Google cloud mirror.  But it looks like a specfic issue with the fields or dimensions being inconsistent between two experiments or something?  Difficult to be sure without seeing the full output.

**Comment by maritsandstad** [comment]
@normansteinert: have you tested which models have this issue? This issue could be a reason to bump the inbuilt caching thing up the priority list, possibly including a "check-availibility-for-models-not-in-cache"-option, which would then get you a results which isn't necessarily  always the same, but always progressively more or equally comprehensive as what you had before. Also, is it just the datagetter definition line that throws this error. It's not, right? It's the subsequent looping over models? How do you get the models to loop over? The datagetter should in principle know which models have available data, but if you rerun only parts of a notebook that may be a problem...

**Comment by normansteinert** [comment]
Yes, that's correct. The data getter works fine and just gets whatever it finds. It seems to be just a handful of models that do not work, so you will still end up with a good 35 or so of models. The error then comes from creating the training data (with make_meteor_training_data). I took out the models that didn't work manually so far. To have these models or not does not make a big difference for the multi-model results. But it may require reworking the list of models that work/don't work depending on when you run the script and do the caching.
I think a functionality might be nice that can check the data before it is turned into METEOR training data, so you don't have to worry about error messages when running the script, even though the exact list of models might vary slightly. I also think that this is slightly separate from the caching, as this is just a means to run the script with loads of models. AndI have the latter in my script and it is also in one of Ben's example notebooks.

**Comment by maritsandstad** [comment]
It's been a minute, but at least at some point the datagetter could get you a list of available models. That list was what I got the model list from earlier (rather than making a manual one). You could always cross check that list with some list you had made for yourself to get the models that you want that there is data for. I can double check that that's still available next week though. @benmsanderson made some changes to make it more flexible, but maybe that also got rid of that functionality, though I shouldn't think so...

**Comment by normansteinert** [comment]
There might be some confusion here. I am using the model list that the data getter generates. So, it should always be the list of models that it thinks are available. From that perspective, getting an inconsistent list every time you run the script is affecting reproducibility a bit, but in my experience the list doesn't vary significantly (and it always results in 30+ models, which should be fine), and it should also make the script run successfully each time (just with a slightly different list of models).
Where the problems come in is 1) when creating the training data ('make_meteor_training_data') (either in-script or saving to cache), 2) when creating the patterns, and 3) when calling the data for example for plotting:
  - for point (1), although they are listed by the data getter, some models give the 'x not in tuple' message. I found that if you do this often enough (over days), you will eventually overcome this issue and it gets data for all the models. So this hints at a data availability issue. 
  - for point (2), even if you managed to successfully obtain all training data, some models fail when producing the patterns ('MeteorPatternScaling'), giving a 'NaN values detected' error message I attach here below. When creating the data getter list with the following experiments: "piControl", "abrupt-4xCO2", "historical", "ssp126", "ssp245", "ssp370", "ssp585", it results in 35 models listed, from which IPSL-CM6A-LR and TaiESM1 for example create the 'NaN' error message.
  - for point (3), in which 'EC-Earth3' gives a ''float' object is not subscriptable' error message calling the data when plotting. 

![Screenshot 2024-11-25 at 09 07 49](https://github.com/user-attachments/assets/79bf5027-bfdd-4d4b-9df7-d8385ff79498)

![Screenshot 2024-11-25 at 09 03 37](https://github.com/user-attachments/assets/24d0517b-8ab1-4d72-bbd0-6a9db465085c)

**Comment by maritsandstad** [comment]
@normansteinert have you pushed a notebook/script something that produces these errors to a branch where I can see it?

**Comment by normansteinert** [comment]
@maritsandstad I just pushed the latest version of the plotting script to the 'description_paper_figures' branch. You should be able to reproduce the errors from there.
